src/main/java/com/endava/internship/collections/StudentMap.java

```diff
@@ -1,78 +1,344 @@
 package com.endava.internship.collections;
 
-import java.util.Collection;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class StudentMap<K extends Comparable<K>, V> implements Map<K, V> {
+    
+    static class Entry<K, V> implements Map.Entry<K, V> {
+
+        private final K key;
+        private V value;
+
+        private Entry<K, V> left;
+        private Entry<K, V> right;
+        private Entry<K, V> parent;
+
+        public Entry(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public K getKey() {
+            return key;
+        }
+
+        @Override
+        public V getValue() {
+            return value;
+        }
+
+        @Override
+        public V setValue(V value) {
+            V oldValue = this.value;
+            this.value = value;
+            return oldValue;
+        }
+
+        @Override
+        public String toString() {
+            return "{" + key + " = " + value + "}\n";
+        }
+    }
+
+    private Entry<K, V> root;
+    private int size;
+    private final Comparator<? super K> comparator;
+
+    public StudentMap() {
+        size = 0;
+        comparator = null;
+    }
+
+    public StudentMap(Comparator<? super  K> comparator) {
+        size = 0;
+        this.comparator = comparator;
+    }
+
+    private int compare(K key1, K key2) {
+        if(key1 == null && key2 == null) {
+            return 0;
+        } else {
+            if (key1 == null) {
+                return -1;
+            } else if (key2 == null) {
+                return 1;
+            }
+        }
+
+        if(comparator == null) {
+            Comparable<? super K> comparableKey = (Comparable<? super K>) key1;
+            return comparableKey.compareTo(key2);
+        } else {
+            return comparator.compare(key1, key2);
+        }
+    }
+
+    private Entry<K, V> search(Object key) {
+        Entry<K, V> currentNode = root;
+        int compareResult = compare((K) key, currentNode.key);
+        try {
+            if(compareResult != 0) {
+                do {
+                    compareResult = compare((K) key, currentNode.key);
+                    if(compareResult > 0) {
+                        if(currentNode.right == null) {
+                            return null;
+                        }
+                        currentNode = currentNode.right;
+                    } else
+                    if(compareResult < 0) {
+                        if(currentNode.left == null) {
+                            return null;
+                        }
+                        currentNode = currentNode.left;
+                    }
+                } while(compareResult != 0);
```

Comment:
Do we really need to check the same condition twice?
compareResult != 0
May be exists way to check it only once? It will reduce cognitive complexity of code, less code - less bugs, easier to read and fix it.